gcc/rust/parse/rust-parse-impl.h

```diff
@@ -7210,315 +7210,46 @@ Parser<ManagedTokenSource>::parse_method ()
-   *  - tuple index expr      expr '.' tuple_index
-   *  - struct expr           path_in_expr (and optional other stuff)
-   *  - enum variant expr     path_in_expr (and optional other stuff)
-   *      this means that there is no syntactic difference between an enum
-   * variant and a struct
-   *      - only name resolution can tell the difference. Thus, maybe rework
-   * AST to take this into account ("struct or enum" nodes?)
-   *  - (function) call expr  expr '(' call_params? ')'
-   *  - method call expr      expr '.' path_expr_segment '(' call_params? ')'
-   *  - field expr            expr '.' identifier
-   *      note that method call expr is preferred, i.e. field expr must not be
-   * followed by parenthesised expression sequence.
-   *  - closure expr          'move'? ( '||' | '|' closure_params? '|' ) (
-   * expr | '->' type_no_bounds block_expr )
-   *  - continue expr         'continue' labelled_lifetime?
-   *  - break expr            'break' labelled_lifetime? expr?
-   *  - range expr            many different types but all involve '..' or
-   * '..='
-   *  - return expr           'return' as 1st tok
-   *  - macro invocation      identifier then :: or identifier then !
-   * (simple_path '!')
-   *
-   * any that have rules beginning with 'expr' should probably be
-   * pratt-parsed,
-   * with parsing type to use determined by token AND lookahead. */
-
-  // ok well at least can do easy ones
-  const_TokenPtr t = lexer.peek_token ();
-  switch (t->get_id ())
+  if (restrictions.consume_semi)
     {
-    case RETURN_TOK:
-      // return expr
-      return parse_return_expr (std::move (outer_attrs));
-    case BREAK:
-      // break expr
-      return parse_break_expr (std::move (outer_attrs));
-    case CONTINUE:
-      // continue expr
-      return parse_continue_expr (std::move (outer_attrs));
-    case MOVE:
-      // closure expr (though not all closure exprs require this)
-      return parse_closure_expr (std::move (outer_attrs));
-      default: {
-	/* HACK: piggyback on pratt parsed expr and abuse polymorphism to
-	 * essentially downcast */
-
-	std::unique_ptr<AST::Expr> expr
-	  = parse_expr (std::move (outer_attrs), restrictions);
-
-	if (expr == nullptr)
-	  {
-	    Error error (t->get_locus (),
-			 "failed to parse expression for expression without "
-			 "block (pratt-parsed expression is null)");
-	    add_error (std::move (error));
-
-	    return nullptr;
-	  }
-
-	std::unique_ptr<AST::ExprWithoutBlock> expr_without_block (
-	  expr->as_expr_without_block ());
-
-	if (expr_without_block != nullptr)
-	  {
-	    return expr_without_block;
-	  }
-	else
-	  {
-	    Error error (t->get_locus (),
-			 "converted expr without block is null");
-	    add_error (std::move (error));
-
-	    return nullptr;
-	  }
-      }
+      if (skip_token (SEMICOLON))
+	{
+	  has_semi = true;
+	}
+      else if (expr->is_expr_without_block ())
+	{
+	  return nullptr;
+	}
```

Comment:
same here but feel free to ignore these nitpicks
```suggestion
      if (skip_token (SEMICOLON))
	  has_semi = true;
      else if (expr->is_expr_without_block ())
	  return nullptr;
```